crates/sui-core/src/narwhal_manager/mod.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use arc_swap::ArcSwap;
+use fastcrypto::bls12381;
+use fastcrypto::traits::KeyPair;
+use futures::future::join_all;
+use narwhal_config::{Committee, Parameters, SharedWorkerCache, WorkerId};
+use narwhal_executor::ExecutionState;
+use narwhal_node::{Node, NodeStorage};
+use narwhal_types::ReconfigureNotification;
+use narwhal_worker::TransactionValidator;
+use prometheus::Registry;
+use std::path::PathBuf;
+use std::sync::Arc;
+use sui_types::crypto::NetworkKeyPair;
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::task::JoinHandle;
+
+pub struct NarwhalManager {
+    pub join_handle: JoinHandle<()>,
+    pub tx_start: Sender<(Arc<Committee>, Registry)>,
+    pub tx_stop: Sender<()>,
+}
+
+pub struct NarwhalConfiguration<
+    State: ExecutionState + Send + Sync + 'static,
+    TxValidator: TransactionValidator,
+> {
+    pub primary_keypair: bls12381::min_sig::BLS12381KeyPair,
+    pub network_keypair: NetworkKeyPair,
+    pub worker_ids_and_keypairs: Vec<(WorkerId, NetworkKeyPair)>,
+    pub worker_cache: SharedWorkerCache,
+
+    pub storage_base_path: PathBuf,
+    pub parameters: Parameters,
+    pub execution_state: Arc<State>,
+    pub tx_validator: TxValidator,
+}
+
+pub async fn run_narwhal_manager<State, TxValidator>(
+    config: NarwhalConfiguration<State, TxValidator>,
+    mut tr_start: Receiver<(Arc<Committee>, Registry)>,
+    mut tr_stop: Receiver<()>,
+) where
+    State: ExecutionState + Send + Sync + 'static,
+    TxValidator: TransactionValidator,
+{
+    let port = config
+        .parameters
+        .network_admin_server
+        .primary_network_admin_server_port;
+
+    loop {
+        // Copy the config for this iteration of the loop
+        let mut id_keypair_copy = Vec::new();
+        for (id, keypair) in &config.worker_ids_and_keypairs {
+            id_keypair_copy.push((id.clone(), keypair.copy()));
+        }
+        let config_copy = NarwhalConfiguration {
+            primary_keypair: config.primary_keypair.copy(),
+            network_keypair: config.network_keypair.copy(),
+            worker_ids_and_keypairs: id_keypair_copy,
+            worker_cache: config.worker_cache.clone(),
+            storage_base_path: config.storage_base_path.clone(),
+            parameters: config.parameters.clone(),
+            execution_state: config.execution_state.clone(),
+            tx_validator: config.tx_validator.clone(),
+        };
```

Comment:
I believe those will be provided on the next `start` but since we don't have the on-chain narwhal config we just use the same info, right?